File: prob1.py
import numpy as np
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.datasets import fetch_20newsgroups
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from scipy.sparse import csr_matrix

def load_data():
    # Load the 20 Newsgroups dataset
    newsgroups = fetch_20newsgroups(subset='all', remove=('headers', 'footers', 'quotes'))
    X, y = newsgroups.data, newsgroups.target
    return X, y, newsgroups.target_names

def preprocess_data(X_train, X_test, min_occurrences=1000):
    # Convert documents to bag-of-words representation
    vectorizer = CountVectorizer(max_features=60000)
    X_train_counts = vectorizer.fit_transform(X_train)
    X_test_counts = vectorizer.transform(X_test)
    
    # Compute word frequencies in training set
    word_counts = np.array(X_train_counts.sum(axis=0)).flatten()
    frequent_words = np.where(word_counts > min_occurrences)[0]
    
    # Reduce feature space by keeping only frequent words
    X_train_pruned = X_train_counts[:, frequent_words]
    X_test_pruned = X_test_counts[:, frequent_words]
    
    return X_train_pruned, X_test_pruned

def estimate_probabilities(X_train, y_train, num_classes):
    # Compute class priors P(y=c)
    class_counts = np.bincount(y_train)
    pi_c = class_counts / class_counts.sum()
    
    # Compute P(x|y) (multinomial word probabilities per class)
    vocab_size = X_train.shape[1]
    pc = np.zeros((num_classes, vocab_size))
    
    for c in range(num_classes):
        class_docs = X_train[y_train == c]
        word_counts = np.array(class_docs.sum(axis=0)).flatten()
        pc[c] = (word_counts + 1) / (word_counts.sum() + vocab_size)  # Laplace smoothing
    
    return pc, pi_c

def train_multinomial_nb(X_train, y_train, num_classes):
    pc, pi_c = estimate_probabilities(X_train, y_train, num_classes)
    wc = np.log(pc)  # Compute log probabilities
    beta_c = np.log(pi_c)
    return wc, beta_c

def predict(X, wc, beta_c):
    # Compute log-probability for each class
    log_probs = X @ wc.T + beta_c
    return np.argmax(log_probs, axis=1)

# Load and split dataset
X, y, class_names = load_data()
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Preprocess data
X_train_pruned, X_test_pruned = preprocess_data(X_train, X_test)

# Train Naïve Bayes classifier
num_classes = len(class_names)
wc, beta_c = train_multinomial_nb(X_train_pruned, y_train, num_classes)

# Make predictions
y_pred = predict(X_test_pruned, wc, beta_c)

# Compute accuracy
accuracy = np.mean(y_pred == y_test)
print(f"Prediction Accuracy: {accuracy * 100:.2f}%")

File: prob1.jl
using MLJ, TextAnalysis, SparseArrays, Downloads, DataStructures, StatsBase
using Statistics, Random, LinearAlgebra

function download_and_extract_data()
    url = "http://qwone.com/~jason/20Newsgroups/20news-bydate.tar.gz"
    filename = "20news-bydate.tar.gz"
    foldername = "20news-bydate"
    
    if !isdir(foldername)
        println("Downloading dataset...")
        Downloads.download(url, filename)
        run(`tar -xzf $filename`)
        println("Dataset extracted.")
    end
    return foldername
end

function load_data()
    foldername = download_and_extract_data()
    train_folder = joinpath(foldername, "20news-bydate-train")
    test_folder = joinpath(foldername, "20news-bydate-test")
    
    categories = readdir(train_folder)
    X_train, y_train = [], []
    X_test, y_test = [], []
    
    for (label, category) in enumerate(categories)
        train_category_path = joinpath(train_folder, category)
        test_category_path = joinpath(test_folder, category)
        
        for file in readdir(train_category_path)
            push!(X_train, String(read(joinpath(train_category_path, file))))
            push!(y_train, label)
        end
        
        for file in readdir(test_category_path)
            push!(X_test, String(read(joinpath(test_category_path, file))))
            push!(y_test, label)
        end
    end
    
    return X_train, X_test, y_train, y_test, categories
end

function preprocess_data(X_train, X_test, min_occurrences=1000)
    word_counts = DefaultDict{String, Int}(0)
    
    function tokenize_texts(texts)
        tokens_list = Vector{Vector{String}}(undef, length(texts))
        Threads.@threads for i in eachindex(texts)
            tokens_list[i] = split(texts[i], r"\\s+")
        end
        return tokens_list
    end
    
    X_train_tokens = tokenize_texts(X_train)
    X_test_tokens = tokenize_texts(X_test)
    
    for tokens in X_train_tokens
        for token in tokens
            word_counts[token] += 1
        end
    end
    
    frequent_words = Set([word for (word, count) in word_counts if count > min_occurrences])
    word_index = Dict(word => i for (i, word) in enumerate(frequent_words))
    
    function vectorize(tokens)
        vec = spzeros(length(frequent_words))
        for token in tokens
            if haskey(word_index, token)
                vec[word_index[token]] += 1
            end
        end
        return vec
    end
    
    X_train_pruned = sparse(hcat([vectorize(tokens) for tokens in X_train_tokens]...)')
    X_test_pruned = sparse(hcat([vectorize(tokens) for tokens in X_test_tokens]...)')
    
    return X_train_pruned, X_test_pruned
end

function estimate_probabilities(X_train, y_train, num_classes)
    class_counts = countmap(y_train)
    total_count = sum(values(class_counts))
    pi_c = Dict(k => v / total_count for (k, v) in class_counts)
    
    vocab_size = size(X_train, 2)
    pc = Dict(c => zeros(vocab_size) for c in 1:num_classes)
    
    for i in 1:size(X_train, 1)
        c = y_train[i]
        pc[c] .+= X_train[i, :]
    end
    
    for c in keys(pc)
        pc[c] = (pc[c] .+ 1) / (sum(pc[c]) + vocab_size)
    end
    
    return pc, pi_c
end

function train_multinomial_nb(X_train, y_train, num_classes)
    pc, pi_c = estimate_probabilities(X_train, y_train, num_classes)
    wc = Dict(c => log.(pc[c]) for c in keys(pc))
    beta_c = Dict(c => log(pi_c[c]) for c in keys(pi_c))
    return wc, beta_c
end

function predict(X, wc, beta_c)
    num_classes = length(wc)
    log_probs = hcat([X * wc[c] .+ beta_c[c] for c in 1:num_classes]...)
    return argmax(log_probs, dims=2)[:]
end

X_train, X_test, y_train, y_test, class_names = load_data()
X_train_pruned, X_test_pruned = preprocess_data(X_train, X_test)
num_classes = length(class_names)
w_c, beta_c = train_multinomial_nb(X_train_pruned, y_train, num_classes)
y_pred = predict(X_test_pruned, w_c, beta_c)
accuracy = mean(y_pred .== y_test)
println("Prediction Accuracy: ", round(accuracy * 100, digits=2), "%")
